core/src/main/java/com/linecorp/armeria/common/stream/EventLoopStreamMessage.java

```diff
@@ -0,0 +1,428 @@
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.CommonPools;
+import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.common.RequestContext;
+
+import io.netty.channel.EventLoop;
+
+/**
+ * A {@link StreamMessage} optimized for when writes and reads all happen on the provided {@link EventLoop},
+ * removing the need for most synchronization from the hot path. This should satisfy standard cases when using
+ * Armeria. It is not required for writes or reads to use the provided {@link EventLoop} but in such a case, it
+ * will be significantly faster to use {@link DefaultStreamMessage}.
+ *
+ * <p>Note that when {@link Subscription#cancel()} or {@link #abort()} are called from a different thread, the
+ * stream will continue to signal objects until demand is satisfied, rather than stopping in the middle. If this
+ * is an issue, it is recommended to use {@link DefaultStreamMessage}.
+ */
+// NB: Methods in this class prefixed with 'do' must be run on the stream's event loop.
+public class EventLoopStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<EventLoopStreamMessage> abortedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(EventLoopStreamMessage.class, "aborted");
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<EventLoopStreamMessage> subscribedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(EventLoopStreamMessage.class, "subscribed");
+
+    private static final Logger logger = LoggerFactory.getLogger(EventLoopStreamMessage.class);
+
+    private final EventLoop eventLoop;
+    private final Queue<Object> queue;
+
+    private SubscriptionImpl subscription;
+    private long demand;
+    private boolean invokedOnSubscribe;
+    private boolean inOnNext;
+
+    private State state = State.OPEN;
+
+    @SuppressWarnings("unused")
+    private volatile int subscribed;  // set only via subscribedUpdater
+    @SuppressWarnings("unused")
+    private volatile int aborted;  // set only via abortedUpdater
+    private volatile boolean isOpen = true;
+    private volatile boolean wroteAny;
+
+    /**
+     * Creates a new {@link EventLoopStreamMessage} which executes all writes on an arbitrary {@link EventLoop}.
+     * It is highly recommended to use {@link #EventLoopStreamMessage(EventLoop)} instead to allow writes to
+     * happen on the same {@link EventLoop} as this stream's.
+     */
+    public EventLoopStreamMessage() {
+        this(RequestContext.mapCurrent(RequestContext::eventLoop, () -> {
+            logger.debug(
+                    "Creating EventLoopStreamMessage without specifying EventLoop. This will be very slow.");
```

Comment:
- will be very slow if subscriber runs in a different EventLoop?
- This can produce a lot of log messages. Could you log it for the first-time only? We could keep the stack trace of problematic calls and log only when the warning message was not logged before at the same call stack. (Just like Netty's leak detector suppresses duplicate log messages for leaks)

Reply:
Hmm - not very experienced with recognizing stack traces, let me know if my approach works :)